File: src/services/serverOdsa.ts
/**
 * Server-Initiated ODSA Handler (ap2011)
 *
 * Handles server-initiated eSIM lifecycle operations,
 * typically scoped to enterprise/managed device contexts.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { EntitlementStatus, SubscriptionResult } from '../protocol/statusCodes.js';
import { getActivationCode } from './mockSmdp.js';
import type { OdsaContext } from './odsaCommon.js';
import { buildOdsaBaseConfig } from './odsaCommon.js';

export interface ServerOdsaConfigData {
  enterpriseId?: string;
  smdpAddress?: string;
  profileType?: string;
  subscriptionState?: 'eligible' | 'pending' | 'active' | 'suspended';
  serviceFlowUrl?: string;
  serviceFlowUserData?: string;
}

export function buildServerOdsaConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
  odsaContext?: OdsaContext,
): ApplicationConfig {
  const data = (configData ?? {}) as ServerOdsaConfigData;
  const operation = odsaContext?.operation;

  if (!operation) {
    return {
      appId: 'ap2011',
      entitlementStatus: status,
      provStatus,
      tcStatus,
    };
  }

  switch (operation) {
    case 'CheckEligibility':
      return handleCheckEligibility(status, provStatus, tcStatus, data);
    case 'ManageSubscription':
      return handleManageSubscription(status, provStatus, tcStatus, data);
    case 'ManageService':
      return handleManageService(status, provStatus, tcStatus, data);
    default:
      return {
        appId: 'ap2011',
        entitlementStatus: status,
        provStatus,
        tcStatus,
      };
  }
}

function handleCheckEligibility(
  status: number, provStatus: number, tcStatus: number, data: ServerOdsaConfigData,
): ApplicationConfig {
  const isEligible = data.subscriptionState === 'eligible' || data.subscriptionState === 'active';
  const entitlementStatus = isEligible ? EntitlementStatus.ENABLED : EntitlementStatus.DISABLED;
  const extra: Record<string, string> = {};
  if (data.enterpriseId) extra.EnterpriseId = data.enterpriseId;
  return buildOdsaBaseConfig('ap2011', entitlementStatus, provStatus, tcStatus, SubscriptionResult.DONE, extra);
}

function handleManageSubscription(
  status: number, provStatus: number, tcStatus: number, data: ServerOdsaConfigData,
): ApplicationConfig {
  if (data.subscriptionState === 'eligible' && data.serviceFlowUrl) {
    const extra: Record<string, string> = { ServiceFlow_URL: data.serviceFlowUrl };
    if (data.serviceFlowUserData) extra.ServiceFlow_UserData = data.serviceFlowUserData;
    if (data.enterpriseId) extra.EnterpriseId = data.enterpriseId;
    return buildOdsaBaseConfig('ap2011', status, provStatus, tcStatus, SubscriptionResult.CONTINUE_TO_WS, extra);
  }

  if (data.subscriptionState === 'active' && data.smdpAddress) {
    const profileKey = data.profileType ?? 'default';
    const activation = getActivationCode(profileKey);
    return buildOdsaBaseConfig('ap2011', status, provStatus, tcStatus, SubscriptionResult.DOWNLOAD_PROFILE, {
      'SMDP+Address': data.smdpAddress,
      'SMDP+ActivationCode': activation.activationCode,
      ProfileICCID: activation.iccid,
    });
  }

  return buildOdsaBaseConfig('ap2011', status, provStatus, tcStatus, SubscriptionResult.DONE);
}

function handleManageService(
  status: number, provStatus: number, tcStatus: number, _data: ServerOdsaConfigData,
): ApplicationConfig {
  return buildOdsaBaseConfig('ap2011', status, provStatus, tcStatus, SubscriptionResult.DONE);
}

File: internal/services/server_odsa.go
package services

import (
	"github.com/hkassaei/entitlements-as-a-service-poc-go/internal/protocol"
)

// ServerOdsaConfigData holds configData for ap2011.
type ServerOdsaConfigData struct {
	EnterpriseID        string `json:"enterpriseId,omitempty"`
	SMDPAddress         string `json:"smdpAddress,omitempty"`
	ProfileType         string `json:"profileType,omitempty"`
	SubscriptionState   string `json:"subscriptionState,omitempty"`
	ServiceFlowURL      string `json:"serviceFlowUrl,omitempty"`
	ServiceFlowUserData string `json:"serviceFlowUserData,omitempty"`
}

// BuildServerOdsaConfig builds the ApplicationConfig for Server-Initiated ODSA (ap2011).
func BuildServerOdsaConfig(status, provStatus, tcStatus int, configData *ServerOdsaConfigData, ctx *OdsaContext) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &ServerOdsaConfigData{}
	}

	operation := ""
	if ctx != nil {
		operation = ctx.Operation
	}

	if operation == "" {
		return &protocol.ApplicationConfig{
			AppID:             "ap2011",
			EntitlementStatus: status,
			ProvStatus:        protocol.IntPtr(provStatus),
			TcStatus:          protocol.IntPtr(tcStatus),
		}
	}

	switch operation {
	case protocol.OpCheckEligibility:
		return serverOdsaCheckEligibility(provStatus, tcStatus, configData)
	case protocol.OpManageSubscription:
		return serverOdsaManageSubscription(status, provStatus, tcStatus, configData)
	case protocol.OpManageService:
		return BuildOdsaBaseConfig("ap2011", status, provStatus, tcStatus, protocol.SubscriptionResultDone, nil)
	default:
		return &protocol.ApplicationConfig{
			AppID:             "ap2011",
			EntitlementStatus: status,
			ProvStatus:        protocol.IntPtr(provStatus),
			TcStatus:          protocol.IntPtr(tcStatus),
		}
	}
}

func serverOdsaCheckEligibility(provStatus, tcStatus int, data *ServerOdsaConfigData) *protocol.ApplicationConfig {
	isEligible := data.SubscriptionState == "eligible" || data.SubscriptionState == "active"
	entitlementStatus := protocol.EntitlementStatusDisabled
	if isEligible {
		entitlementStatus = protocol.EntitlementStatusEnabled
	}
	extra := map[string]string{}
	if data.EnterpriseID != "" {
		extra["EnterpriseId"] = data.EnterpriseID
	}
	return BuildOdsaBaseConfig("ap2011", entitlementStatus, provStatus, tcStatus, protocol.SubscriptionResultDone, extra)
}

func serverOdsaManageSubscription(status, provStatus, tcStatus int, data *ServerOdsaConfigData) *protocol.ApplicationConfig {
	if data.SubscriptionState == "eligible" && data.ServiceFlowURL != "" {
		extra := map[string]string{"ServiceFlow_URL": data.ServiceFlowURL}
		if data.ServiceFlowUserData != "" {
			extra["ServiceFlow_UserData"] = data.ServiceFlowUserData
		}
		if data.EnterpriseID != "" {
			extra["EnterpriseId"] = data.EnterpriseID
		}
		return BuildOdsaBaseConfig("ap2011", status, provStatus, tcStatus, protocol.SubscriptionResultContinueToWS, extra)
	}

	if data.SubscriptionState == "active" && data.SMDPAddress != "" {
		profileKey := data.ProfileType
		if profileKey == "" {
			profileKey = "default"
		}
		activation := GetActivationCode(profileKey)
		return BuildOdsaBaseConfig("ap2011", status, provStatus, tcStatus, protocol.SubscriptionResultDownloadProfile, map[string]string{
			"SMDP+Address":        data.SMDPAddress,
			"SMDP+ActivationCode": activation.ActivationCode,
			"ProfileICCID":        activation.ICCID,
		})
	}

	return BuildOdsaBaseConfig("ap2011", status, provStatus, tcStatus, protocol.SubscriptionResultDone, nil)
}
